packages/block-library/src/table/state.js

```diff
@@ -276,16 +276,25 @@ export function deleteColumn( state, {
  * @return {Object} New table state.
  */
 export function toggleSection( state, sectionName ) {
-	// Section exists, replace it with an empty row to remove it.
+	const sectionSaved = `${ sectionName }Saved`;
+
+	// Section exists, save it and replace it with an empty row to remove it.
 	if ( ! isEmptyTableSection( state[ sectionName ] ) ) {
-		return { [ sectionName ]: [] };
+		return { [ sectionName ]: [], [ sectionSaved ]: [ ...state[ sectionName ] ] };
+	}
+
+	// Check if we already have this section
+	if ( state[ sectionSaved ] ) {
+		return {
+			[ sectionName ]: [ ...state[ sectionSaved ] ],
```

Comment:
It might be good to also clear `sectionSaved` state when the data is restored. At the moment the issue following happens:

1. Add a table
2. Toggle on header rows
3. Add some text to the header
4. Use the 'Delete row' option to delete the header.
5. Toggle the header back on

Expected Result: The header text shouldn't have been restored since it was deleted.
Actual Result: The header text was restored.